app/http.php

```diff
@@ -0,0 +1,366 @@
+            $executorStates->set($executor, [
+                'status' => 'online',
+                'hostname' => $executor,
+                'state' =>  \json_encode([])
+            ]);
+        }
+    }
+});
+
+Console::log('State of executors at startup:');
+
+go(function () use ($register) {
+    $executorStates = $register->get('executorStates');
+
+    $executors = \explode(',', (string) App::getEnv('OPEN_RUNTIMES_PROXY_EXECUTORS', ''));
+
+    foreach ($executors as $executor) {
+        $state = $executorStates->exists($executor) ? $executorStates->get($executor) : null;
+
+        if ($state === null) {
+            Console::warning('Executor ' . $executor . ' has unknown state.');
+        } else {
+            Console::log('Executor ' . $executor . ' is ' . ($state['status'] ?? 'unknown') . '.');
+        }
+    }
+});
+
+Swoole\Event::wait();
+
+// Wildcard action
+App::error()
+    ->inject('request')
+    ->inject('response')
+    ->inject('roundRobinAtomic')
+    ->inject('executorStates')
+    ->action(function (Request $request, Response $response, Atomic $roundRobinAtomic, Table $executorStates) {
+        $secretKey = \explode(' ', $request->getHeader('authorization', ''))[1] ?? '';
+
+        if (empty($secretKey)) {
+            throw new Exception('Incorrect proxy key.', 401);
+        }
+        if ($secretKey !== App::getEnv('OPEN_RUNTIMES_PROXY_SECRET', '')) {
+            throw new Exception('Incorrect proxy key.', 401);
+        }
+
+        $roundRobinIndex = $roundRobinAtomic->get();
+
+        $algoType = App::getEnv('OPEN_RUNTIMES_PROXY_ALGORITHM', '');
+        $algo = match ($algoType) {
+            'round-robin' => new RoundRobin($roundRobinIndex - 1), // Atomic indexes from 1. Balancing library indexes from 0. That's why -1
+            'random' => new Random(),
+            default => new Random()
+        };
+
+        $balancing = new Balancing($algo);
+
+        $balancing->addFilter(fn ($option) => $option->getState('status', 'offline') === 'online');
+
+        foreach ($executorStates as $state) {
+            $balancing->addOption(new Option($state));
+        }
+
+        $body = \json_decode($request->getRawPayload() ? $request->getRawPayload() : '{}', true);
+        $runtimeId = $body['runtimeId'] ?? null;
+        // TODO: @Meldiron Use RuntimeID in CPU-based adapter
+
+        $option = $balancing->run();
+
+        if ($option === null) {
+            throw new Exception('No online executor found', 404);
+        }
+
+        if ($algo instanceof RoundRobin) {
+            $roundRobinAtomic->cmpset($roundRobinIndex, $algo->getIndex() + 1); // +1 because of -1 earlier above
+        }
+
+        $executor = [];
+
+        foreach ($option->getStateKeys() as $key) {
+            $executor[$key] = $option->getState($key);
+        }
+
+        Console::success('Executing on ' . $executor['hostname']);
+
+        $client = new Client($executor['hostname'], 80);
```

Comment:
I also just realized this is the Client we use for tests. We should probably relocate it to something like `../src/Proxy/HTTP/Client.php` and add some tests for it. We do have future plans to have a dedicated Utopia client we could reuse across tests and other places instead of CURL.

Reply:
This client comes from `use Swoole\Coroutine\Http\Client;`.
It's a different client from the one in tests.

I would like to use this one here as it's recommended (Swoole stuff must work best with Swoole).
I would like to prevent using it in tests as tests run in PHPUtil, most likely without proper Swoole support. I am already getting warnings about coroutines when unit-testing Health class.

Reply:
Swoole client is fine. Once we release our own client we should make sure it is the best option for us.